### PR TITLE
Login screen

### PR DESCRIPTION
**Pull request by Gotalicp**
- Remade ui of login screen in compose
- Fixed viewmodel for compose

### CODE REVIEW

File: app/src/main/java/bg/zahov/app/ui/authentication/AuthenticationState.kt [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package bg.zahov.app.ui.authentication
+
+sealed interface AuthenticationState {
+    data object Authenticate : AuthenticationState
+    data class Default(val uiInfo: UiInfo, val stateCounter: Int) : AuthenticationState
+    data class Notify(val uiInfo: UiInfo, val message: String, val stateCounter: Int = 0) :
+        AuthenticationState
+}
+
+data class UiInfo(
```
Comment (dimitarstoyanovtumba) [inline app/src/main/java/bg/zahov/app/ui/authentication/AuthenticationState.kt:10]: Please add kdoc to all public classes and functions.

File: app/src/main/java/bg/zahov/app/ui/authentication/login/LoginFragment.kt [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+
+    override fun onCreateView(
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?,
+    ): View {
+        return ComposeView(requireContext()).apply {
+            setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
+            setContent {
+                LoginScreen(loginViewModel, findNavController())
```
Comment (dimitarstoyanovtumba) [inline app/src/main/java/bg/zahov/app/ui/authentication/login/LoginFragment.kt:26]: The login screen itself doesn't need to hold control over the navController. Use a lambda as an argument callback instead and call the navigation function from this fragment.

File: app/src/main/java/bg/zahov/app/ui/authentication/login/LoginFragment.kt [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+
+class LoginFragment : Fragment() {
+    private val loginViewModel: LoginViewModel by viewModels()
+
+    override fun onCreateView(
+        inflater: LayoutInflater,
+        container: ViewGroup?,
+        savedInstanceState: Bundle?,
+    ): View {
+        return ComposeView(requireContext()).apply {
```
Comment (dimitarstoyanovtumba) [inline app/src/main/java/bg/zahov/app/ui/authentication/login/LoginFragment.kt:23]: requireContext and requireActivity can return a null or throw an exception. Use `context` instead.

File: app/src/main/java/bg/zahov/app/ui/authentication/login/LoginScreen.kt [hunk 4]
```diff
@@ -0,0 +1,143 @@
...
+        ).show()
+    }
+    val interactionSource = remember { MutableInteractionSource() }
+
+    Column(
+        modifier = Modifier
+            .fillMaxSize()
+    ) {
+        Text(
+            "Login",
```
Comment (dimitarstoyanovtumba) [inline app/src/main/java/bg/zahov/app/ui/authentication/login/LoginScreen.kt:62]: Please extract all string resources.

File: app/src/main/java/bg/zahov/app/ui/authentication/login/LoginScreen.kt [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+            modifier = Modifier
+                .align(Alignment.CenterHorizontally)
+                .padding(top = 130.dp)
+        )
+        Column(
+            modifier = Modifier
+                .width(240.dp)
+                .padding(top = 20.dp)
+                .align(Alignment.CenterHorizontally),
+            verticalArrangement = Arrangement.spacedBy(15.dp)
```
Comment (dimitarstoyanovtumba) [inline app/src/main/java/bg/zahov/app/ui/authentication/login/LoginScreen.kt:74]: nit: As a good practice use paddings that are divisible by 4.

File: app/src/main/java/bg/zahov/app/ui/authentication/login/LoginViewModel.kt [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+import bg.zahov.app.ui.authentication.UiInfo
+import bg.zahov.app.util.isEmail
+import kotlinx.coroutines.launch
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import java.lang.IllegalArgumentException
+
+class LoginViewModel(application: Application) : AndroidViewModel(application) {
+    private val info: UiInfo = UiInfo()
```
Comment (dimitarstoyanovtumba) [inline app/src/main/java/bg/zahov/app/ui/authentication/login/LoginViewModel.kt:18]: Observable pattern with Kotlin Flow should be used here instead.

File: app/src/main/java/bg/zahov/app/ui/authentication/signup/SignupScreen.kt [hunk 7]
```diff
@@ -0,0 +1,143 @@
...
+        Text(
+            modifier = Modifier.padding(bottom = 20.dp),
+            text = "Sign up",
+            style = MaterialTheme.typography.headlineLarge,
+            textAlign = TextAlign.Center,
+            color = Color.White,
+            fontWeight = FontWeight.Bold
+        )
+        Column(verticalArrangement = Arrangement.spacedBy(15.dp)) {
+//            CommonTextField(
```
Comment (dimitarstoyanovtumba) [inline app/src/main/java/bg/zahov/app/ui/authentication/signup/SignupScreen.kt:78]: Let's avoid committing commented code.

File: app/src/main/java/bg/zahov/app/ui/authentication/signup/SignupScreen.kt [hunk 8]
```diff
@@ -0,0 +1,143 @@
...
+        ),
+            onClick = {
+                navController.navigate(R.id.signup_to_login)
+            }) {
+            Text(text = stringResource(R.string.already_have_account_text), color = Color.White)
+        }
+    }
+}
+
+fun showToast(message: String, context: Context) {
```
Comment (dimitarstoyanovtumba) [inline app/src/main/java/bg/zahov/app/ui/authentication/signup/SignupScreen.kt:141]: This function seems redundant.

File: app/src/main/java/bg/zahov/app/ui/custom/CommonPasswordField.kt [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+        )
+    },
+    shape: Shape = RoundedCornerShape(10.dp),
+    colors: TextFieldColors = TextFieldDefaults.colors(
+        focusedIndicatorColor = Color.Transparent,
+        unfocusedIndicatorColor = Color.Transparent
+    ),
+    onPasswordChange: (String) -> Unit,
+    onPasswordVisibilityChange: () -> Unit,
+    @SuppressLint("ModifierParameter") modifier: Modifier = Modifier
```
Comment (dimitarstoyanovtumba) [inline app/src/main/java/bg/zahov/app/ui/custom/CommonPasswordField.kt:44]: Better to follow the lint rules than to suppress them.

File: app/src/main/java/bg/zahov/app/ui/authentication/login/LoginViewModel.kt [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+        }
+
+        viewModelScope.launch {
+            auth.passwordResetByEmail(email)
+                .addOnSuccessListener { notifyChange("Rest password email has been sent") }
+                .addOnFailureListener { notifyChange("Couldn't send reset password email!") }
+        }
+    }
+
+    private fun notifyChange(text: String? = null) {
```
[the remark this answers is not in the document]
Comment (Gotalicp) [reply app/src/main/java/bg/zahov/app/ui/authentication/login/LoginViewModel.kt:132]: waiting for Spas :)

**Review by trifonov-i — CHANGES_REQUESTED** [review]

**Pull request merged**